5주차 고양이 장난감가게 만들기 - 심화

humblEgo opened the pull request:
안녕하세요~! 
이번 주도 수요일부터 가열차게 할 수 있을 것 같아요. 잘 부탁드립니다!

## 과제 목표

- [x] 테스트 커버리지 100% 달성
- [x] API 테스트 통과
- [x] E2E 테스트 통과

## 6/1 현재 추가로 작업할 내용 체크리스트
- [x] UserController 테스트 작성
- [x] Product 관련 테스트 구조화
- [x] 모든 public 메서드에 대해 javadoc 작성

humblEgo committed:
web 의존성 파일을 설치하라

humblEgo committed:
test 의존성 파일을 설치하라

humblEgo committed:
Dozer mapper를 추가하라

humblEgo committed:
Dozer Mapper 객체를 추가하고, 필요없는 메서드를 삭제하라

humblEgo committed:
User 클래스를 테스트하라

humblEgo committed:
User와 UserDto를 추가하라

humblEgo committed:
UserRepository, JpaUserRepository를 추가하라

humblEgo committed:
UserServiceTest를 추가하라

humblEgo committed:
UserService를 구현하라

humblEgo committed:
ProductController에 `@CrossOrigin`를 추가하라

humblEgo committed:
NotFoundUserException을 추가하라

humblEgo committed:
UserController를 구현하라

humblEgo committed:
잘못된 '/users/{id}' PATCH 테스트 케이스를 수정하라

존재하지 않는 유저의 id를 path 변수로 요청하고, 404 상태
코드를 응답 받고자 하는 테스트가 잘못 작성되어있었습니다.
PATCH 테스트에 email을 요청 body에 포함시키지
않아서 UserDto 유효성 검사를 통과하지 못하도록 테스트케이스가
구성되어 있어서, 400 상태코드를 응답 받게 되어있었기 때문입니다.
테스트 요청 body에 email을 포함시켜서 해결했습니다.

humblEgo committed:
createUser 테스트시 verify 대상을 변경하라

verify 대상이 sampleUser 객체이면
createUser 메소드 내에서 새로운 객체를 만들어서 save하므로
통과할 수 없습니다. 이에 verify 통과를 위해 변경했습니다.

humblEgo committed:
updateUser시 존재하지 않는 유저 id가 주어진 경우를 추가하라

humblEgo committed:
updateUser에서 던지는 예외의 종류를 변경하라

던져진 예외의 종류를 명확히 표현하기 위해,
존재하지 않는 유저 id가 주어진 경우,
EmptyResultDataAccessException이 아니라
NotFoundUserException을 던지도록 수정했습니다.

humblEgo committed:
AppTest를 추가하라

테스트 커버리지 100%를 위함

humblEgo committed:
테스트를 구조화하고, toString 테스트를 추가하라

humblEgo committed:
NotFoundException을 orElseThrow로 던지도록 수정하라

humblEgo committed:
쓰이지 않는 import문을 제거하라

humblEgo committed:
PATCH '/users/{id}' 테스트에서 email을 삭제하라

humblEgo committed:
User Create, Update시 각기 다른 조건으로 유효성 검사하라

humblEgo committed:
findById 목킹을 고쳐라

humblEgo committed:
UserControllerTest를 추가하라

johngrib commented on app/build.gradle:
```diff
@@ -51,6 +51,9 @@ dependencies {
     compileOnly 'org.projectlombok:lombok:1.18.16'
     annotationProcessor 'org.projectlombok:lombok:1.18.16'
 
+    // Dozer
+    implementation 'com.github.dozermapper:dozer-core:6.4.0'
```
코드 리뷰의 형식으로 DozerMapper의 장단점과 도입 계기 등을 설명해 보세요.

humblEgo replied:
### DozerMapper란?
> Dozer is a Java Bean to Java Bean mapper that recursively copies data from one object to another, it is an open source mapping framework that is robust, generic, flexible, reusable, and configurable.

한 객체에서 다른 객체로 데이터를 재귀적으로 복사하는 오픈소스 맵핑 프레임워크입니다. 특히 `layered architecture`에서 layer간 데이터를 전달하기 위해 DTO를 만들 때 필요한 데이터를 맵핑하는데 도움을 줍니다.

### 도입계기
Entity 객체의 attribute 정보를 DTO로, 혹은 DTO attribute 정보를 Entity 객체로 데이터를 옮길 때, getter/setter를 활용해서 일일이 맵핑해야 했습니다. DozerMapper는 이 맵핑 작업 공수를 없애기 위해 도입했습니다.

장점
 - 객체간 데이터 매핑을 쉽게 할 수 있습니다.
 - reflection을 사용하기 때문에, bytecode generation을 활용한 맵핑 프레임워크에 비해 일정한 퍼포먼스를 냅니다.
 
단점
 - runtime 시점에 reflection을 통해 맵핑을 하기 때문에 맵핑할 객체 사이즈가 커질 수록 성능 이슈가 생길 수 있습니다. 
 - custom mapping을 위해서는 XML 파일을 따로 작성해야합니다.

### 참고
 - [참고: 맵핑 프레임워크간 퍼포먼스 비교](https://www.baeldung.com/java-performance-mapping-frameworks)
 - [Dozer FAQ](http://dozer.sourceforge.net/documentation/faq.html) : 궁금하신 점 있으면 댓글 남겨주시거나 이 링크 참고해주시면 감사하겠습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,41 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class UserService {
```
`@Service`, `@Controller`와 같은 주요 클래스에는 JavaDoc 주석을 달아주세요. `public` 메소드에도 전부 달아주시고요.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/User.java:
```diff
@@ -0,0 +1,31 @@
+@Entity
+@Builder
+@Getter
+@NoArgsConstructor
+@AllArgsConstructor
```
이 두 애노테이션을 사용할 때 주의할 점이 무엇이 있는지 조사해 보세요. 좋은 공부가 될 거에요.

humblEgo replied:
네, 둘 모두 [Lombok](https://projectlombok.org/) 라이브러리에서 제공하는 어노테이션이며, 생성자 작성 공수를 줄여주지만 각각 주의할 점이 있습니다.

### `@NoArgsConstructor`

#### 기능
파라미터가 없는 생성자를 생성합니다.

#### 주의할 점

1. 필드들이 final로 생성되어 있는 경우에는 필드를 초기화 할 수 없기 때문에 생성자를 만들 수 없고, 에러가 발생합니다. 해결책으로는 `force=true` 옵션을 이용해서 강제로 초기화하는 것을 고려할 수 있습니다.
2. `@Nonnull`과 같이 필드에 제약조건이 설정되어 있는 경우, 추후 초기화를 진행하기 전까지 생성자 내 null-check 로직이 생성되지 않습니다.
3. `@java.beans.ConstructorProperties`를 추가할 순 없습니다.

### `@AllArgsConstructor`

#### 주의할 점

1. 파라미터 생성 순서가 필드 선언 순서에 의존합니다. 만약 필드 선언 순서를 바꿨다면 파라미터와 필드가 적절하게 맵핑되었는지 확인할 필요가 있습니다.
2. 이미 final 로 초기화된 필드가 있다면, 이는 `@AllArgsConstructor`로 만들어지는 생성자의 파라미터로 포함되지 않습니다.
3. 그런데 또 final로 선언되었지만 초기화되지 않은 필드가 있다면, 이는 `@AllArgsConstructor`로 만들어지는 생성자의 파라미터로 포함됩니다.

### 참고
- https://projectlombok.org/features/constructor
- http://stevenmwesigwa.com/tutorials/project-lombok/8/how-to-use-allargsconstructor-annotation-with-project-lombok-in-java-applicat
- https://velog.io/@rosa/Lombok-%EC%A7%80%EC%96%91%ED%95%B4%EC%95%BC-%ED%95%A0-annotation

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,215 @@
+        @DisplayName("유저 정보가 주어지면")
+        class Context_of_valid_user {
+
+            private UserDto givenUserDto;
+
+            @BeforeEach
+            void setup() {
+                givenUserDto = sampleUserDto;
+            }
```
`it` 절이 하나 뿐이어서 이 코드는 `@BeforeEach` 없이 한 줄로 표현해도 괜찮겠어요.

humblEgo replied:
아하 넵 Context를 좀 더 명확히 보여주고 싶어서 이렇게 표현해봤는데, 조금 형식적인 느낌이 들긴 했어요ㅎㅎ
더 깔끔해지겠네요!

별개로 `createUser`에 추가로 테스트되어야 할 것이 있는지, 어디까지 테스트를 해야할지 조금 고민되더라구요.
지금 생각나는건, `UserDto` 객체 대신 `null`을 인자로 넘겨주는 케이스 정도만 생각나는 상태입니다.

1) 추가로 테스트되어야할 케이스가 있을까요?
2) 추가로 테스트되어야할 케이스가 있는지 체크하는 노하우가 있을까요? 시야를 넓히고 싶네용..

johngrib replied:
숫자를 입력하는 메소드가 있다고 칩시다. 이 메소드에서 테스트해야 하는 값들은 다음과 같습니다.

- 개발자가 상상한 정상 입력 범위의 숫자들.
- 정상 입력 범위의 특수 값, 경계값들: `0`, `Integer.MAX`, `Integer.MIN`, `-1`, `255` 등등
- 정상 입력 범위를 아득히 넘어서는 큰 수 또는 작은 수.
- `null`
- 경우에 따라 다를 수 있지만, `char`

자 `createUser`에서는 무엇을 얼마나 더 테스트하면 좋을까요? 테스트코드는 소프트웨어의 신뢰성을 확보하기 위한 도구이므로 많으면 많을수록 좋지만, 너무 많아도 관리하기가 힘들어집니다.

humblEgo replied:
답변 감사합니다! 예시를 보니 좀 더 감이 생기는 것 같습니다.

`createUser`에는 아래 두 경우만 테스트해도 충분하다고 생각했습니다.
  - `UserDto` 속성 값이 정상적인 경우
  - `null`

왜냐면 `UserDto` 속성 값이 정상적이지 않거나 특수한 값(ex 한글)이 들어간 경우는 `userControllerWeb` 테스트와 `UserDto` 테스트에서 확인하는게 맞다고 생각했기 때문입니다.

johngrib reviewed (commented):
`ProductContollerTest`에 보면 다음과 같이 deprecated 된 상수를 사용하는 코드가 있습니다. 해당 상수의 JavaDoc을 읽고 JavaDoc이 시키는대로 조치해 보세요.

![image](https://user-images.githubusercontent.com/1855714/120326716-fc5a9a80-c323-11eb-9ab8-f203acbb8339.png)

humblEgo commented:
안녕하세요 종립님!
죄송합니다, 오늘 학습을 제대로 진행하지 못했네요ㅠㅠ 주신 조언 모두 확인했습니다, 내일까지 학습 & 반영해보겠습니다!

humblEgo committed:
Deprecreated된 상수를 수정하라

'MediaType.APPLICATION_JSON_UTF8'가 deprecated된 것을 확인하여 권장되는 다른
상수인 'MediaType.APPLICATION_JSON_VALUE'로 수정하였습니다.

humblEgo commented:
> `ProductContollerTest`에 보면 다음과 같이 deprecated 된 상수를 사용하는 코드가 있습니다. 해당 상수의 JavaDoc을 읽고 JavaDoc이 시키는대로 조치해 보세요.
> 
> ![image](https://user-images.githubusercontent.com/1855714/120326716-fc5a9a80-c323-11eb-9ab8-f203acbb8339.png)


d8d5a15

아래 내용 감안하여 위 커밋에 반영했습니다! 
> A String equivalent of APPLICATION_JSON_UTF8.
Deprecated
as of 5.2 in favor of APPLICATION_JSON_VALUE since major browsers like Chrome now comply with the specification  and interpret correctly UTF-8 special characters without requiring a charset=UTF-8 parameter.

humblEgo committed:
UserController에 javadoc을 추가하라

humblEgo committed:
NotFoundErrorAdvice에 javadoc을 추가하라

humblEgo committed:
UserService에 javadoc을 추가하라

humblEgo committed:
ProductService에 javadoc을 추가하라

humblEgo committed:
HelloController에 javadoc을 추가하라

humblEgo committed:
ProductController에 javadoc을 추가하라

humblEgo committed:
Product에 javadoc을 추가하라

humblEgo committed:
User에 javadoc을 추가하라

humblEgo committed:
ProductNotFoundException의 javadoc을 추가하라

humblEgo committed:
NotFoundUserException의javadoc을 추가하라

humblEgo committed:
App에 javadoc을 추가하라

humblEgo committed:
ErrorResponse에 javadoc을 추가하라

johngrib commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,15 +1,32 @@
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.Bean;
 
+/**
+ * 스프링 부트 앱
+ */
```
이 주석을 통해 우리가 알 수 있는 것이 없는 것 같아요.

humblEgo replied:
주석을 위한 주석이었던 것 같습니다.

아래 두 선택지를 생각해봤고, 2)번을 택하여 `앱 메인 클래스`라고 적어봤습니다, 간단한듯 어렵네요.
1) 제거
2) 메인문이 포함된 클래스임을 표시

johngrib commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,15 +1,32 @@
-    }
-
+    /**
+     * 스프링 부트 애플리케이션을 시동합니다.
```
지금 개발하고 있는 애플리케이션이 스프링 부트를 사용해 만든 애플리케이션이긴 합니다. 그런데 이 애플리케이션을 부르는 다른 방법이 없을까요?

humblEgo replied:
서버 역할이 드러났으면 해서, `서버를 시동합니다.` 로 변경해봤습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,15 +1,32 @@
     }
+
+    /**
+     * DozerMapper 객체를 Java Bean으로 만듭니다.
```
이렇게 할 수 있겠죠.
```suggestion
     * Mapper Bean을 리턴합니다.
```
```suggestion
     * Mapper 구현체를 리턴합니다.
```
```suggestion
     * Mapper Bean을 정의합니다.
```

humblEgo replied:
주석에 너무 상세하게 내용을 드러냈더니 `DozerMapper`에 의존성이 생겨버렸었군요..!

추천 주신 것 중 3번째가 가장 취향에 맞는데요, 반환값도 명시해보았습니다.
`Mapper Bean을 정의하고 리턴합니다.`

johngrib replied:
상황을 빠르고 정확히 파악하시는 것 같아 기쁩니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/ProductNotFoundException.java:
```diff
@@ -1,6 +1,14 @@
 package com.codesoom.assignment;
 
+/**
+ * 상품을 찾지 못한 경우의 예외 객체
```
예외는 던지는 것이죠.
```suggestion
 * 상품을 찾지 못했을 때 던집니다.
```

humblEgo replied:
아하 무얼하는 녀석인지 좀 더 잘 드러나는 것 같습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/ProductNotFoundException.java:
```diff
@@ -1,6 +1,14 @@
+ */
 public class ProductNotFoundException extends RuntimeException {
+    /**
+     * 찾지 못한 상품의 id를 조합한 메시지를 포함하여, 예외 객체를 생성합니다.
```
어떤 생각이 드나요?
```suggestion
     * 생성자.
```

humblEgo replied:
후아 제 주석이 넘 장황했던 것 같습니다.

예외 클래스 내부에 위치해있으므로, `생성자`라고만 적어도 충분히 제가 전하고자 했던 정보가 전달되네요.

이런게 누적되면 생산성으로도 이어질 것 같습니다. 
읽는 사람 입장에서는 읽어야하는 공수가 늘고, 쓰는 사람 입장에서도 쓰는데 시간이 걸리고..

이미 충분히 의미가 통하는 단어가 있다면 센스있게 적용하려 노력해야겠습니다!

johngrib replied:
사실 `생성자` 쯤 되면 그냥 생략해도 괜찮다고 생각해요 ㅎㅎ

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -9,23 +9,48 @@
 import java.util.List;
 
+/**
+ * 상품 데이터에 대한 요청을 처리합니다.
```
이 주석을 통해 이 서비스가 다루는 대상이 얼마나 포괄적인지를 생각해 볼 수 있어요. 지금은 기능이 몇 개 안 되지만 `상품 데이터`를 처리하는 요청의 종류가 100개를 넘어가게 된다면 어떻게 해야 좋을까요?

humblEgo replied:
지금 상태라면 이 서비스가 다루는 대상이 너무 포괄적이어서 모든 요청을 포함하게 될 것 같습니다.
그렇다면 유지보수성이 떨어질 것이구요.
결국 요청의 응집도를 고려해서 클래스를 나눠야할 것 같습니다. 
이 때 `상품 데이터에 대한 요청을 처리합니다.`라는 javadoc을 각 클래스의 역할에 맞게 수정해야할 것 같습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -9,23 +9,48 @@
     private final ProductRepository productRepository;
 
+    /**
+     * 상품 저장소 객체를 주입 받습니다.
```
상품 저장소 객체를 주입받는 것이 이 메소드(생성자)의 목적인가요?

humblEgo replied:
서비스 객체를 만드는 것이 목적입니다.

객체를 주입 받기 위해 생성자에 인자를 정의했으니, 이를 드러내야 한다고 생각했었어요.
`생성자.`로 깔끔하게 적어버리는게 좋을 것 같습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -0,0 +1,72 @@
+     * @param id 갱신할 유저 id
+     * @param source 갱신할 내용
+     * @return 갱신한 유저 정보
+     * @throws NotFoundUserException
+     *      유저를 찾지 못했을 때 던지는 예외
```
```suggestion
     * @throws NotFoundUserException 유저를 찾지 못했을 경우
```

humblEgo replied:
throws NotFoundUserException 을 포함해서 한 눈에 슥- 읽히고, 라인 수도 줄어들어서 훨씬 좋네요!

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/HelloController.java:
```diff
@@ -3,6 +3,9 @@
 import org.springframework.web.bind.annotation.RestController;
 
+/**
+ * ROOT URL의 요청을 처리합니다.
```
만약 이 컨트롤러가 담당하는 URL이 `/`에서 `/hello`로 바뀌게 된다면 이 주석도 바뀌어야 하지 않을까요? 이 컨트롤러가 진짜로 담당하고 있는 것은 `/`인가요? 아니면 `hello` 인가요?

humblEgo replied:
`hello` 입니다. 말씀대로 URL을 특정하면서 주석을 적기에는, URL 장체가 너무 자주 바뀔 가능성이 있는 녀석으로 보이네요.
`Hello API의 요청을 처리합니다.` 로 변경해봤습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/NotFoundErrorAdvice.java:
```diff
@@ -2,18 +2,41 @@
 import org.springframework.web.bind.annotation.ResponseStatus;
 
+/**
+ * 컨트롤러 layer에서 던지는 예외를 처리합니다.
```
클래스 주석에서는 `컨트롤러 layer에서 던지는 예외를 처리합니다.`라고 되어있는데요, 클래스 이름과 괴리감이 있어요. 클래스 이름을 보면 컨트롤러 레이어이건, 서비스 레이어이건 `무언가 찾지 못했을 경우`의 예외 처리를 하는 것처럼 보입니다.

humblEgo replied:
아 정말 그러네요! 클래스명 짓듯 좀 더 신경써서 주석을 적어야겠습니다.

`무언가 찾지 못했을 경우 던져진 예외를 처리합니다.`로 수정해보았습니다!

johngrib replied:
이제는 `무언가`가 문제가 되겠죠.

humblEgo replied:
맞아요, 음.. `무언가`를 대체할 말로 `요청 받은 자원` 정도가 떠오릅니다.

humblEgo committed:
App Javadoc을 수정하라

humblEgo committed:
UserDto가 null인 경우의 테스트를 추가하라

humblEgo committed:
ProductNotFoundException의 javadoc을 수정하라

humblEgo committed:
HelloController javadoc을 수정하라

humblEgo committed:
ProductService javadoc을 수정하라

humblEgo committed:
UserService javadoc을 수정하라

humblEgo committed:
NotFoundErrorAdvice javadoc을 수정하라

humblEgo committed:
테스트에서 한글이 깨지는 에러를 해결하라

APPLICATION_JSON_VALUE로 수정했다가 한글이 깨지는
현상이 있었습니다. 이에 UTF-8 인코딩 필터와 응답 인코딩
설정을 추가했습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/App.java:
```diff
@@ -1,15 +1,32 @@
+import org.springframework.context.annotation.Bean;
 
+/**
+ * 애플리케이션 메인 클래스
```
```suggestion
 * 애플리케이션 메인 클래스.
```

humblEgo replied:
앗 dot(.)을 빼먹었네요, 전수체크해서 수정 해보겠습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/ProductController.java:
```diff
@@ -9,31 +10,68 @@
 public class ProductController {
     private final ProductService productService;
 
+    /**
+     * 상품 서비스 객체를 주입 받습니다.
+     *
+     * @param productService 상품 서비스 객체
+     */
```
불필요한 내용이 있습니다.

humblEgo replied:
전수조사하여, 비슷한 패턴을 모두 `생성자.`로 수정했습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/ProductController.java:
```diff
@@ -9,31 +10,68 @@
     }
 
+    /**
+     * 200 상태 코드와 모든 상품 목록을 응답합니다.
```
`200 OK`를 응답한다는 사실은 생략해도 괜찮겠죠.

humblEgo replied:
아하, 하긴 default 값 느낌이긴하네요.
`200 OK` 응답한다는 사실은 생략하고, 그 외 상태코드 응답은 모두 javadoc에 명시해봤습니다!

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Product.java:
```diff
@@ -9,6 +9,9 @@
 import javax.persistence.GeneratedValue;
 import javax.persistence.Id;
 
+/**
+ * 상품 엔티티
+ */
```
단순하게 갈수록 좋습니다.
```suggestion
/**
 * 상품.
 */
```

johngrib commented on app/src/main/java/com/codesoom/assignment/exception/NotFoundUserException.java:
```diff
@@ -0,0 +1,15 @@
+package com.codesoom.assignment.exception;
+
+/**
+ * 유저를 찾지 못한 경우 던지는 예외 객체
```
```suggestion
 * 유저를 찾지 못한 경우 던집니다.
```

humblEgo replied:
앗 놓쳤었네요; 전체 체크해서 수정해보겠습니다~!

johngrib reviewed (commented):
이제 JavaDoc은 너무 잘 작성하셔서 따로 코멘트해드릴 게 거의 없네요!

humblEgo committed:
javadoc에 dot(.)을 추가하라

humblEgo committed:
예외는 던지는 것임을 javadoc에 드러내라

humblEgo committed:
ErrorResponse javadoc을 수정하라

humblEgo committed:
UserControllerWebTest를 추가하라

humblEgo committed:
필요 없는 `@toString` 어노테이션을 제거하라

humblEgo committed:
ProductTest에 builder 케이스를 추가하고 가독성을 높여라

humblEgo committed:
ProductDataTest를 추가하라

humblEgo committed:
UserDtoTest를 추가하라

humblEgo committed:
쓰지 않는 import문을 삭제하라

humblEgo committed:
javaDoc을 간략하게 수정하라

불필요한 내용, 200 OK 응답한다는 사실, 좀 더 단순화할 수 있는 문구를 삭제 처리했습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/dto/UserDto.java:
```diff
@@ -0,0 +1,35 @@
+@Getter
+@Setter
+@Builder
+@NoArgsConstructor
```
이건 취향 차이라고도 할 수 있을텐데요, 이 애노테이션은 안 쓰는 게 더 좋다고 생각합니다. 그냥 기본 생성자를 하나 만들면 되니까요. 이건 제 취향이니 이 리뷰는 무시하셔도 좋습니다.

johngrib commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,240 @@
+        given(userRepository.findAll())
+                .willReturn(List.of(sampleUser));
+
+        given(userRepository.findById(EXISTENT_ID))
+                .willReturn(Optional.ofNullable(sampleUser));
+
+        given(userRepository.findById(NON_EXISTENT_ID))
+                .willReturn(Optional.empty());
+
+        willThrow(EmptyResultDataAccessException.class)
+                .given(userRepository).deleteById(NON_EXISTENT_ID);
```
이 세 **전제**에 대해서는 테스트에서의 중요도를 생각해 보면 따로 분리해 강조하는 것도 괜찮겠어요.

humblEgo replied:
메소드로 분리하고 의미가 드러나게 네이밍하면 더 좋겠다는 생각이 드네요!

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -0,0 +1,224 @@
+    @DisplayName("create 메소드는")
+    class Describe_of_create {
+
+        @Nested
+        @DisplayName("유저 정보가 주어지면")
+        class Context_of_valid_user_dto {
+
```
유저 정보를 여기에 선언해 주세요.

humblEgo replied:
e9b8c7d

반영해보았습니다!

johngrib commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -0,0 +1,224 @@
+            @Test
+            @DisplayName("유저를 삭제한다")
+            void it_deletes_user() {
+                userController.destroy(givenId);
```
지금 보니 새삼 무시무시한 메소드 이름이네요. `destroy`와 `delete`는 어감이 꽤 달라서 생소한 느낌이 듭니다.

humblEgo replied:
`destroy`는 좀 더 강한 어감이 있다고 생각합니다. 
다음엔 내부에서 호출되는 메소드의 이름과의 일관성도 맞춰 봐야겠습니다. 그 책임만큼의 동작이 이뤄진다는게 느껴지게요.

johngrib commented on app/src/test/java/com/codesoom/assignment/dto/ProductDataTest.java:
```diff
@@ -0,0 +1,107 @@
+        productData = builder.build();
+    }
+
+    @Nested
+    @DisplayName("getId 메소드는")
+    class Describe_of_getId {
+
+        @Test
+        @DisplayName("id를 반환한다")
+        void it_returns_id() {
+            assertThat(productData.getId()).isEqualTo(defaultId);
+        }
+    }
```
커버리지를 위한 테스트이지만 꽤나 지루한 느낌이 들어 답답하군요. 고생하셨습니다.

humblEgo replied:
커버리지를 위한 테스트여서 차라리 `D-C-I` 패턴을 적용하지 않고 가볍게 테스트해버렸다면 덜 지루했을 것 같긴 합니다.

그간 좋은 피드백 주셔서 감사합니다! :)

johngrib reviewed (commented):
한 주간 고생 많으셨습니다. 주간 회고를 작성해 공유해 주세요.

humblEgo committed:
컨텍스트에 맞게 유저 정보를 선언하라
